### PR TITLE
[CI] Better support for overlaying submodules

### PR DESCRIPTION
**Pull request by JasonGross**
The overlay infrastructure now supports declaring that the root directory of some builds is a subfolder of a given repository.  This allows overlay-writers to override the repository and branch (and override the subfolder with "") to indicate that the repository should be cloned from elsewhere.

This should hopefully provide a relatively seamless experience for overlay writers who need to update submodules of projects like fiat-crypto.  In service of that, we split off
ci-{coqutil,kami,riscv_coq,rupicola}

### CODE REVIEW

**Comment by JasonGross** [comment]
@coqbot run full ci

(@coq/coqbot-maintainers why does coqbot not automatically run full CI if I create the PR with the `request: full CI` label?)

**Review by proux01 — CHANGES_REQUESTED** [review]
The split into multiple CI targets is probably a good thing to better understand the structure.
I'm less fond of the complexification of the CI machinery. What about the following alternative:
- have some `coq-master` (or whatever name you prefer) branch in bedrock2 (/ coq-util / ...) that will be used by Coq CI / receive overlay PRs
- use whatever mechanism you want to update the fiat-crypto submodule to the coq-master branch of bedrock2 each time it receives an overlay

You already have to update the submodule anyway, it's just that the information of which version of bedrock2 to use in the CI would be stored by the coq-master branch of bedrock2 (and just be mirrored by the submodule). From Coq's CI point of view, that would make submodules almost transparent (except all the EXTERNAL... arguments) and it would be much easier.

**Comment by JasonGross** [comment]
This is not as easy as you make it sound.  We'd have to duplicate the CI infrastructure (quadratically in the depth of the dependency chain) to ensure that we don't merge any breaking changes into those branches.  Whereas currently we can just rely on dependabot and local project CIs.

@andres-erbsen do you have ideas or thoughts?

I agree that the complexity here is not ideal, but it allows submodule dependencies to be mostly transparent across all projects for constant complexity overhead.

**Comment by andres-erbsen** [comment]
I must be misunderstanding, as what @proux01 proposes sounds like the *current* state of affairs to me (where `coq-master` is called `master` and "whatever mechanism" is dependabot). This PR is to enable Coq CI to pass using an overlay before the bedrock2 submodule in-fiat crypto is updated by dependabot (or a Coq developer). FWIW, I don't like the CI complexity either and would prefer if all these projects were a monorepo, but alas. Evaluating the concept behind this PR on its own, I am in favor on the basis that the state after this PR would be more consistent with what I think Coq developers expect from CI/overlay infrastructure than the state before.

**Comment by JasonGross** [comment]
> I must be misunderstanding, as what @proux01 proposes sounds like the _current_ state of affairs to me

We don't maintain the contract that, e.g., bedrock2 master always is compatible with fiat-crypto master.  If we did, we could split out the dependencies and point them at the master branches of the corresponding repos.  This PR is to enable the combination of:
- submitting overlays without waiting for a submodule bump merge
- not enforcing the strict contract that {bedrock2,coqutil,rupicola,fiat-crypto,coq-riscv,kami} master must always all be compatible
- ensuring that Coq's CI does not randomly break because of incompatibilities in the tip of master of {bedrock2,coqutil,rupicola,fiat-crypto,coq-riscv,kami}

File: dev/ci/user-overlays/README.md
Comment (SkySkimmer) [inline dev/ci/user-overlays/README.md:76]: This sounds complicated, not sure we want to merge.

Comment (JasonGross) [reply dev/ci/user-overlays/README.md:76]: This is just documenting what is already the case, except that today there is no guidance on how to make overlays.  In fact the situation with this PR is significantly simpler: `_build_ci/<job name>` will be a symlink to the appropriate folder, and this complicated procedure is only needed to determine the branch against which the PR should be made.  (This procedure can also be used to determine the repository, though this can also be done with `git remote -v`)

Maybe the text could make this clearer?

Comment (JasonGross) [reply dev/ci/user-overlays/README.md:76]: @SkySkimmer what do you think of this?
```suggestion
  - If the project in `ci-basic-overlay.sh` includes a `submodule_directory`
    argument (for example, the last argument of `project bedrock2
    "https://github.com/mit-plv/fiat-crypto" "master" "" "rupicola/bedrock2"`),
    you can list all branches the current commit is on with
    `git fetch; git branch -r --contains $(git rev-parse HEAD)`
    (use `HEAD^` instead of `HEAD` if you have already committed), and you can
    see if the parent repository contains a branch specification by executing
    `(cd "$(pwd)/.."; git config -f .gitmodules --get-regexp '^submodule\..*\.branch')`
    from the `_build_ci/<job name>` directory.
```

**Comment by ejgallego** [comment]
I need more time to look at this, but I'm principle I feel good about going the direction @JasonGross proposes.

**Comment by JasonGross** [comment]
> I'm not quite sure what it would be -- would we want bedrock2 CI to build fiat-crypto with overlays like Coq CI does for plugins?

I think the most hands-off alternative here is to have a branch ("tested" / "rv32i-tested"?), and every commit to master (but not PRs) test the build with all reverse dependencies and push to the tested branch only if things build.  If we want to avoid duplicating CI infrastructure we can set up all the repos to have opam files and just do `opam pin .` and then `opam install <reverse deps>`.  This would result in quadratic CI overhead if combined with submodules (every project is going to build every project to update the chain of submodules), and also still results in desynchronization when trying to land coupled changes, though.  I don't see a way to deal with coupled changes on CI in this style without overlays.

**Comment by andres-erbsen** [comment]
Are you suggesting we'd also take PRs straight to `tested`? If not, I don't think this satisfies @proux01's request.

**Comment by JasonGross** [comment]
> Are you suggesting we'd also take PRs straight to `tested`? If not, I don't think this satisfies @proux01's request.

Ah, yeah, I missed that part.  You're right, it doesn't satisfy the requirement.

**Comment by proux01** [comment]
> would we want bedrock2 CI to build fiat-crypto with overlays like Coq CI does for plugins?

From Coq point of view, you can manage your CI in every way you want, as long as you maintain the invariants that all branches registered in Coq CI keep building together.

> Ah, yeah, I missed that part. You're right, it doesn't satisfy the requirement.

Sorry, would you mind explaining the use case that appears problematic? I'm a bit lost here.

**Comment by andres-erbsen** [comment]
The way the Bedrock2 maintainers have chosen to manage their CI is to not test against users and instead perform catch-up for non-backwards-compatible changes when bumping the bedrock2 submodule in downstream developments. If we want that to change, it is on us to have a proposal. I may be wearing two hats here, and I do think some change would likely be desirable, but I can't just go to the team and tell them they now have to figure it out. I bring up Coq overlays exactly as an example of how what you are asking for is nontrivial. This Coq PR here is a proposed solution to about the same problem, and it seems much simpler to me. I am open to alternative proposals, but to evaluate them we need enough detail to understand the impact on different contributor workflows (Coq overlays, compatible bedrock2 changes, incompatible bedrock2 changes, etc).

For example, Jason was suggesting that the coq-master branch you wish for would be automatically fast-forwarded from master whenever the invariant you envision is satisfied, but you also want mit-plv/* to accept overlay PRs to the *same* branch, which wouldn't be compatible with fast-forwarding.

**Comment by proux01** [comment]
> The way the Bedrock2 maintainers have chosen to manage their CI is to not test against users

Whereas Coq CI is by definition the exact contrary: testing users. Maybe the two philosophies are intrinsically incompatible?

> This Coq PR here [...] seems much simpler to me.

Well, I certainly understand it feels simpler from your point of view since it transfers, at least part of, the complexity on our side. That's exactly what we are complaining about ;)

> For example, Jason was suggesting that the coq-master branch you wish for would be automatically fast-forwarded from master whenever the invariant you envision is satisfied, but you also want mit-plv/* to accept overlay PRs to the _same_ branch, which wouldn't be compatible with fast-forwarding.

Why so? It's your branch, you do whatever you want with it, as long as the invariant is maintained. Maybe you would need merge or rebase instead of just fast forwarding, but I don't see the issue.

**Comment by andres-erbsen** [comment]
> Well, I certainly understand it feels simpler from your point of view since it transfers, at least part of, the complexity on our side. That's exactly what we are complaining about ;)

No, this is not what I mean. Obviously we should evaluate the overall complexity, combined across Coq, bedrock2, and their users. What I am saying is that I don't see how pushing this problem downstream to bedrock2 would help. In particular, I don't think we should hold up this PR in favor of a solution elsewhere that is yet to be designed.

**Comment by JasonGross** [comment]
> > Well, I certainly understand it feels simpler from your point of view since it transfers, at least part of, the complexity on our side. That's exactly what we are complaining about ;)
> 
> No, this is not what I mean. Obviously we should evaluate the overall complexity, combined across Coq, bedrock2, and their users. What I am saying is that I don't see how pushing this problem downstream to bedrock2 would help.

To be more concrete, this PR does two things:
1. add infrastructure for handling submodules transparently
2. factor out the dependencies of fiat-crypto

The diff-stat of (1) is +33,-1, with an additional +21,-1 of documentation.

To implement [the proposal I made above](https://github.com/coq/coq/pull/18736#issuecomment-1998055947), I estimate a cost of about 300 lines (20 lines of CI + 40 lines of opam files, times 5 repositories).  And it would add 10+ hours of CI time to every update, if we want to keep submodules.

The logic of this PR is not especially complex, and I don't expect it to be particular hard to maintain.  @proux01 Can you be a bit more concrete about the cost that you're concerned about with this PR?

---------

> Why so? It's your branch, you do whatever you want with it, as long as the invariant is maintained. Maybe you would need merge or rebase instead of just fast forwarding, but I don't see the issue.

The issue is that the following invariants seem incompatible:
1. Coq's CI tracks branches
2. The collection of branches that are tracked by Coq's CI always build together / consistently
3. Overlays are made to the branches tracked by Coq's CI
4. Projects have a primary branch where development is done
5. Project developers do not have to manually manage synchronization between the primary development branch and the branch tracked by Coq's CI
6. There is no risk of merge conflicts that have to be manually resolved between the primary development branch and the branch tracked by Coq's CI

In particular, insofar as I can tell, 6 implies
7. Pull requests are only accepted against the primary development branch of the project

which is incompatible with 3.

**Comment by ejgallego** [comment]
Thanks for all the feedback folks, indeed in the CI we took inspiration from the experience some large codebases (Jane Street's OCaml's libs for example) in managing their setup. The way they frame the problem is indeed similar on how you would frame the problem of consistency in a distributed system.

IIUTC, in that case you have to manage autonomy of "branches" vs "reconciliation cost". There is ample evidence that reconciliation cost tends to be exponential in the size of the divergence (in this case delta), so we favor a kind of "greedy" strategy where diverging in CI branches will imply that the people in charge of the CI development will have to bear the reconciliation cost. This has indeed some complex tradeoffs.

I'm unaware of easy solutions to this problem, but indeed, I think we can improve the status quo and evolve our model.

**Comment by proux01** [comment]
> To be more concrete, this PR does two things:
> 
>     1. add infrastructure for handling submodules transparently

Unfortunately, as shown by your own doc, it doesn't really achieve it.

> The issue is that the following invariants seem incompatible:
> 
>     1. Coq's CI tracks branches
> 
>     2. The collection of branches that are tracked by Coq's CI always build together / consistently
> 
>     3. Overlays are made to the branches tracked by Coq's CI
> 
>     4. Projects have a primary branch where development is done
> 
>     5. Project developers do not have to manually manage synchronization between the primary development branch and the branch tracked by Coq's CI
> 
>     6. There is no risk of merge conflicts that have to be manually resolved between the primary development branch and the branch tracked by Coq's CI

Sure but 5. and 6. are in now way requirements of Coq CI. Those are requirements you are adding yourself. Their net effect is to transfer some loadwork from you to us, which seems to be the core of the issue. (BTW, I'm not even sure that 5. is that true, you can probably get it done automatically most of the time)

**Comment by JasonGross** [comment]
> > To be more concrete, this PR does two things:
> > ```
> > 1. add infrastructure for handling submodules transparently
> > ```
> 
> Unfortunately, as shown by your own doc, it doesn't really achieve it.

Are you seriously complaining about the fact the *determining what branch to make the pull request against* can't be read off the overlays file?  And using this complaint to block this PR in a roundabout way rather than simply requesting a trivial fix such as *adding a comment to any line using a submodule* indicating the branch to make the PR against?

**Comment by JasonGross** [comment]
> Sure but 5. and 6. are in now way requirements of Coq CI. Those are requirements you are adding yourself. Their net effect is to transfer some loadwork from you to us, which seems to be the core of the issue

These are reasonable requirements to have of our CI.  Are you being honest about the core issue here, or is the core of your objection that the project this issue is about is part of the bedrock2/fiat-crypto ecosystem?
I think that 30 lines of straightforward code in a part of the system that is independent of Coq itself is a very reasonable cost for the Coq repository to bear to avoid hundreds of lines of code and the ongoing cost of hours of developer time in users of Coq.  It seems to me that Coq frequently adds much more complicated features for much smaller gains in ease of use and I don't generally see you or anybody else objecting to such additions.  I feel frustrated by what seems to me to be biased or deceptively- / incompletely- presented objections.

File: dev/ci/user-overlays/README.md
Comment (SkySkimmer) [reply dev/ci/user-overlays/README.md:76]: TBH I wanted to know what this PR changes in the overlay workflow ("This should hopefully provide a relatively seamless experience for overlay writers" is not very specific) and only saw a bunch of added stuff which was not encouraging.

Comment (SkySkimmer) [reply dev/ci/user-overlays/README.md:76]: I think finding the correct branch should not be made to sound this complex, something like "for projects which are submodules, to determine the branch to open a PR against you should look in the parent project's repository" is probably enough.

Comment (JasonGross) [reply dev/ci/user-overlays/README.md:76]: Sorry about that @SkySkimmer 

For projects that are not submodules / where the overlay does not make changes to any submodules, this PR changes nothing.

For overlays to submodules of projects, the old overlay workflow included additional steps not present in the documentation:
1. cd to the submodule directory
2. make the commit there
3. push to a branch
4. figure out the appropriate repo and branch to make the PR to
5. make the PR
6. make corresponding commits, branches, and PRs to bump the submodule in the parent project (recursively if necessary)
7. write another version of the submodule bump commit that points the submodule at your fork (do this recursively if necessary)
8. write an overlay file

This PR removes all of these steps, making the workflow for writing overlays to submodules identical to the workflow for writing overlays to other projects, with the exception of two details:
1. the repository to make the PR against and the branch name cannot be read off the `ci-basic-overlay` file but must instead be obtained from `git remote -v` and either `git fetch; git branch -r --contains $(git rev-parse HEAD)` or looking at the parent repository
2. when the PR is merged, upstream must merge not only the overlay PR, but also (recursive) submodule bumping PRs (otherwise Coq master will pick up the old versions).  I think part of the contract of using submodules should be that upstream maintainers will not just merge the overlay PRs promptly (as is required of all developments today), but will take care of bumping required submodules promptly on their end, without any additional input from Coq developers.*

As an alternative to (1), we can demand that submodule projects include a comment with this information alongside the project declaration in `ci-basic-overlay`, in which case the only changes for submodule projects are:
1. Determine the upstream repo and branch from the comment rather than the `project` line
2. Tell upstream to take care of merging not just the overlay but also of bumping submodules

*Note that this is a relatively automated process, and so even coq-community projects using submodules would not incur too much cost; we simply tell dependabot to check for updates (go to, e.g., https://github.com/mit-plv/fiat-crypto/network/updates, click "recent update jobs" next to `.submodules`, click "check for updates") and then say `@dependabot squash and merge` on the PRs it creates.

Comment (JasonGross) [reply dev/ci/user-overlays/README.md]: I realized I had not updated `dev/tools/create_overlays.sh`; I added 8 loc to have the generated overlay point to the correct repo and another 8 loc to make it print a comment instructing the developer where to make the PR against.

So now the only differences are:
1. Determine the upstream repo and branch from the end of the project line rather than the middle; or determine them from the comment in the generated overlay file
2. Tell upstream to take care of merging not just the overlay but also of bumping submodules

**Comment by ejgallego** [comment]
> I think that 30 lines of straightforward code in a part of the system that is independent of Coq itself is a very reasonable cost for the Coq repository to bear to avoid hundreds of lines of code and the ongoing cost of hours of developer time in users of Coq.

I fully agree @JasonGross , and indeed let me be clear that your work on this is very much appreciated and I'm sure we understand and address all the challenges we have in managing CI deps and go ahead with this proposal.

**Comment by ejgallego** [comment]
Having a bit more shell doesn't worry me; it is true that adding more shell code is a problem for Coq developers because it is hard to maintain; maybe we could go ahead and start using OCaml to manage this kind of stuff? I dunno.

**Comment by proux01** [comment]
> These are reasonable requirements to have of our CI.

Maybe, I'd just like to *understand* why. Fiat-crypto is pretty much the only thing in CI demanding such things. If it were by far the most elaborate thing in CI that might be understandable, but there are plenty of other elaborate developments in our CI and they all seem to live a happy life without such demands. So without clear detailled explanations, it's hard not to wonder whether you are not just making some potential development process issue on your side drip on us. You might be right that hacking Coq CI rather than changing anything on your dev process might overall be the best solution. But this must come with clear, thourough, explanations of that dev process, and what makes fiat-crypto uniquely require that, so that it's better understood and accepted.

File: dev/ci/ci-common.sh
Comment (SkySkimmer) [inline dev/ci/ci-common.sh:124]: I'm not sure about this -PARENT stuff
I tried `make ci-fiat_crypto DOWNLOAD_ONLY=1` locally and it cloned fiat crypto and its submodules a bunch of times.

Why aren't we cloning fiat_crypto once and symlinking the submodules to it? ie something that would work out to
~~~
cd _build_ci
git clone $fiat_crypto_url fiat_crypto
(cd fiat_crypto && git submodule update --init --recursive)
ln -s  "$(pwd)fiat_crypto/rupicola" rupicola
(etc)
~~~

Comment (JasonGross) [reply dev/ci/ci-common.sh:124]: Sure, I can add another argument to `project` identifying the parent project so we know what folder to clone to

**Review by proux01 — APPROVED** [review]
It's probably not the definitive solution but since it's already an improvement over the current situation, let's move toward merging.

File: dev/ci/ci-common.sh
Comment (JasonGross) [reply dev/ci/ci-common.sh:124]: Hm, this is actually pretty annoying.  We can't update the submodule if the `.git` directory doesn't exist in the parent repo.  The fundamental problem here is that the download logic is inherently local but here you're asking for the logic to be non-local across projects.  Let me see if I can work something out though.  (It's going to make the logic somewhat more complicated though.)

Comment (JasonGross) [reply dev/ci/ci-common.sh:124]: Argh, even https://github.com/coq/coq/pull/18824/files is not sufficient, because the leaf dependencies are going to download the parent project, and the .git will be removed, but then it becomes impossible to reuse the artifact later to adjust the submodules.  I guess I can make it work locally though.

Comment (JasonGross) [reply dev/ci/ci-common.sh]: I've pushed an update that should avoid duplicate downloads when possible, but may rework and simplify the logic here later

Comment (JasonGross) [reply dev/ci/ci-common.sh]: @SkySkimmer I've reorganized the code a bit.  There's now a separate `subproject` function in `ci-basic-overlay.sh` which takes `<name> <parent project> <submodulefolder> <submodule giturl> <submodule branch>` as in `subproject kami fiat_crypto "rupicola/bedrock2/deps/kami" "https://github.com/mit-plv/kami" "rv32i"`.  There is no difference in declaring or submitting overlays, the parent repository should be downloaded only once (unless you have already deleted it's `.git` folder).

I've removed the documentation on submitting overlays to submodules, because I don't think it adds anything anymore, the process is not really different except that you read from the arguments to `subproject` instead of `project`.

**Comment by JasonGross** [comment]
Anyone have any ideas what's up with
```
coqc -R src QuickChick -I plugin QuickChickInterface.v
Error: Can't find file ./QuickChickInterface.v
```
?

**Comment by JasonGross** [comment]
cc  @lemonidas, @Lysxia, @liyishuai, any idea why QuickChick might be failing with
```
coqc -R src QuickChick -I plugin QuickChickInterface.v
Error: Can't find file ./QuickChickInterface.v
```
?  Was there a recent change to master?  Or would removing all files and folders named `.git` somehow result in this?

**Comment by Lysxia** [comment]
Oh yes, I recently changed the build script in QuickChick, and forgot to update it here.

File: dev/ci/ci-common.sh
Comment (SkySkimmer) [inline dev/ci/ci-common.sh:127]: why?

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
This doesn't work locally, I made a bedrock2 overlay, pushed, rm -rf _build_ci, make ci-fiat_crypto DOWNLOAD_ONLY=1

It looks like the overlay was checked out but then the git_download for rupicola reset it.

https://github.com/SkySkimmer/coq/tree/testbr
https://github.com/SkySkimmer/bedrock2/tree/testbr

File: dev/ci/ci-common.sh
Comment (JasonGross) [reply dev/ci/ci-common.sh:127]: Roughly the issue is that downloading the parent does not appropriately overlay unused submodules.  Let me fix that.

**Comment by JasonGross** [comment]
> This doesn't work locally, I made a bedrock2 overlay, pushed, rm -rf _build_ci, make ci-fiat_crypto DOWNLOAD_ONLY=1
> 
> It looks like the overlay was checked out but then the git_download for rupicola reset it.
> 
> https://github.com/SkySkimmer/coq/tree/testbr https://github.com/SkySkimmer/bedrock2/tree/testbr

Thanks, this is fixed now.  The problem was that I declared transitive submodules as subprojects of fiat_crypto, rather than as subprojects of their direct parents.

----

I've also removed the fancy `.git` logic and am instead sharing parent project folders only when the projects are checked out not on the CI.  The CI setup is deeply relying on sibling jobs having disjoint directory structures so that .v timestamps are not scribbled over, and I don't see any way to combine this with downloading repositories in jobs of siblings.

File: dev/ci/ci-common.sh
Comment (JasonGross) [reply dev/ci/ci-common.sh]: I've remove the logic around .git

**Comment by JasonGross** [comment]
CI passes again

**Comment by SkySkimmer** [comment]
Maybe squash a bit, there's several "revert" commits in there

**Review by SkySkimmer — APPROVED** [review]

**Comment by JasonGross** [comment]
I've squashed, there's now only three commits.  Do we need to run the full CI again (I haven't changed any code)?

**Review by ejgallego — APPROVED** [review]
Thanks folks for all the hard work on review and code, this LTGM.

@proux01 , I didn't have the time (yet) to try to understand your concerns with this, but IIUC there is still some remaining items you would like to see improved right?

If so, I'd suggest opening an issue so we document what's left to do / evolve.

**Comment by proux01** [comment]
@ejgallego in the long term maybe (I'm still not entirely sure I have a good understanding of the MIT development process and its compatibility with Coq CI), not in this PR anyway.

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**
